{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36c4cfc8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'passage': 'Aconitine-induced Ca2+ overload causes arrhythmia and triggers apoptosis through p38 MAPK signaling pathway in rats. Aconitine is a major bioactive diterpenoid alkaloid with high content derived from herbal aconitum plants. Emerging evidence indicates that voltage-dependent Na(+) channels have pivotal roles in the cardiotoxicity of aconitine. However, no reports are available on the role of Ca(2+) in aconitine poisoning. In this study, we explored the importance of pathological Ca(2+) signaling in aconitine poisoning in vitro and in vivo. We found that Ca(2+) overload lead to accelerated beating rhythm in adult rat ventricular myocytes and caused arrhythmia in conscious freely moving rats. To investigate effects of aconitine on myocardial injury, we performed cytotoxicity assay in neonatal rat ventricular myocytes (NRVMs), as well as measured lactate dehydrogenase level in the culture medium of NRVMs and activities of serum cardiac enzymes in rats. The results showed that aconitine resulted in myocardial injury and reduced NRVMs viability dose-dependently. To confirm the pro-apoptotic effects, we performed flow cytometric detection, cardiac histology, transmission electron microscopy and terminal deoxynucleotidyl transferase-mediated dUTP-biotin nick end labeling assay. The results showed that aconitine stimulated apoptosis time-dependently. The expression analysis of Ca(2+) handling proteins demonstrated that aconitine promoted Ca(2+) overload through the expression regulation of Ca(2+) handling proteins. The expression analysis of apoptosis-related proteins revealed that pro-apoptotic protein expression was upregulated, and anti-apoptotic protein BCL-2 expression was downregulated. Furthermore, increased phosphorylation of MAPK family members, especially the P-P38/P38 ratio was found in cardiac tissues. Hence, our results suggest that aconitine significantly aggravates Ca(2+) overload and causes arrhythmia and finally promotes apoptotic development via phosphorylation of P38 mitogen-activated protein kinase.',\n",
       " 'entities': [{'id': '0',\n",
       "   'text': 'Aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '1',\n",
       "   'text': 'Ca2+',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '2',\n",
       "   'text': 'arrhythmia',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D001145'},\n",
       "  {'id': '3',\n",
       "   'text': 'p38 MAPK',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '81649'},\n",
       "  {'id': '4', 'text': 'rats', 'type': 'OrganismTaxon', 'identifier': '10116'},\n",
       "  {'id': '5',\n",
       "   'text': 'Aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '6',\n",
       "   'text': 'diterpenoid alkaloid',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': '-'},\n",
       "  {'id': '7',\n",
       "   'text': 'Na(+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D012964'},\n",
       "  {'id': '8',\n",
       "   'text': 'cardiotoxicity',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D066126'},\n",
       "  {'id': '9',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '10',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '11',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '12',\n",
       "   'text': 'poisoning',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D011041'},\n",
       "  {'id': '13',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '14',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '15',\n",
       "   'text': 'poisoning',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D011041'},\n",
       "  {'id': '16',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '17', 'text': 'rat', 'type': 'OrganismTaxon', 'identifier': '10116'},\n",
       "  {'id': '18',\n",
       "   'text': 'arrhythmia',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D001145'},\n",
       "  {'id': '19', 'text': 'rats', 'type': 'OrganismTaxon', 'identifier': '10116'},\n",
       "  {'id': '20',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '21',\n",
       "   'text': 'myocardial injury',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D009202'},\n",
       "  {'id': '22',\n",
       "   'text': 'cytotoxicity',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D064420'},\n",
       "  {'id': '23', 'text': 'rat', 'type': 'OrganismTaxon', 'identifier': '10116'},\n",
       "  {'id': '24',\n",
       "   'text': 'lactate dehydrogenase',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '24533'},\n",
       "  {'id': '25', 'text': 'rats', 'type': 'OrganismTaxon', 'identifier': '10116'},\n",
       "  {'id': '26',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '27',\n",
       "   'text': 'myocardial injury',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D009202'},\n",
       "  {'id': '28', 'text': 'dUTP', 'type': 'ChemicalEntity', 'identifier': '-'},\n",
       "  {'id': '29',\n",
       "   'text': 'biotin',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D001710'},\n",
       "  {'id': '30',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '31',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '32',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '33',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '34',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '35',\n",
       "   'text': 'BCL-2',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '24224'},\n",
       "  {'id': '36',\n",
       "   'text': 'MAPK',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '81649'},\n",
       "  {'id': '37',\n",
       "   'text': 'P38',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '81649'},\n",
       "  {'id': '38',\n",
       "   'text': 'P38',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '81649'},\n",
       "  {'id': '39',\n",
       "   'text': 'aconitine',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D000157'},\n",
       "  {'id': '40',\n",
       "   'text': 'Ca(2+)',\n",
       "   'type': 'ChemicalEntity',\n",
       "   'identifier': 'D002118'},\n",
       "  {'id': '41',\n",
       "   'text': 'arrhythmia',\n",
       "   'type': 'DiseaseOrPhenotypicFeature',\n",
       "   'identifier': 'D001145'},\n",
       "  {'id': '42',\n",
       "   'text': 'P38 mitogen-activated protein kinase',\n",
       "   'type': 'GeneOrGeneProduct',\n",
       "   'identifier': '81649'}],\n",
       " 'relations': [{'entity1': 'D012964',\n",
       "   'entity2': 'D066126',\n",
       "   'relation': 'Association'},\n",
       "  {'entity1': 'D002118',\n",
       "   'entity2': 'D001145',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'D001145', 'entity2': '81649', 'relation': 'Association'},\n",
       "  {'entity1': 'D000157',\n",
       "   'entity2': 'D066126',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'D000157',\n",
       "   'entity2': 'D009202',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'D000157', 'entity2': '81649', 'relation': 'Association'},\n",
       "  {'entity1': 'D000157',\n",
       "   'entity2': 'D002118',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'D000157',\n",
       "   'entity2': 'D001145',\n",
       "   'relation': 'Positive_Correlation'}],\n",
       " 'true_relations': [{'entity1': 'Na(+)',\n",
       "   'entity2': 'cardiotoxicity',\n",
       "   'relation': 'Association'},\n",
       "  {'entity1': 'Ca2+',\n",
       "   'entity2': 'arrhythmia',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'arrhythmia', 'entity2': 'p38 MAPK', 'relation': 'Association'},\n",
       "  {'entity1': 'Aconitine',\n",
       "   'entity2': 'cardiotoxicity',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'Aconitine',\n",
       "   'entity2': 'myocardial injury',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'Aconitine', 'entity2': 'p38 MAPK', 'relation': 'Association'},\n",
       "  {'entity1': 'Aconitine',\n",
       "   'entity2': 'Ca2+',\n",
       "   'relation': 'Positive_Correlation'},\n",
       "  {'entity1': 'Aconitine',\n",
       "   'entity2': 'arrhythmia',\n",
       "   'relation': 'Positive_Correlation'}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Load the JSON data from file\n",
    "file_path = './BioRED/Test.BioC.JSON'  # Replace with your file path\n",
    "with open(file_path, 'r') as file:\n",
    "    data = json.load(file)\n",
    "\n",
    "# Function to combine passages and entities, and format relations\n",
    "def process_documents(documents):\n",
    "    combined_passages = []\n",
    "    combined_entities = []\n",
    "    document_relations = []\n",
    "\n",
    "    for document in documents:\n",
    "        # Combining texts and annotations (entities) for all passages in the document\n",
    "        all_texts = []\n",
    "        all_entities = []\n",
    "        for passage in document['passages']:\n",
    "            all_texts.append(passage['text'])\n",
    "            annotations = passage.get('annotations', [])\n",
    "            for ann in annotations:\n",
    "                entity_info = {\n",
    "                    'id': ann['id'],\n",
    "                    'text': ann['text'],\n",
    "                    'type': ann['infons']['type'],\n",
    "                    'identifier': ann['infons']['identifier']\n",
    "                }\n",
    "                all_entities.append(entity_info)\n",
    "\n",
    "        # Combining all texts into a single passage\n",
    "        combined_passage_text = ' '.join(all_texts)\n",
    "        combined_passages.append(combined_passage_text)\n",
    "        combined_entities.append(all_entities)\n",
    "\n",
    "        # Extracting and formatting relations for the document\n",
    "        relations = []\n",
    "        for relation in document['relations']:\n",
    "            entity1_id = relation['infons']['entity1']\n",
    "            entity2_id = relation['infons']['entity2']\n",
    "            relation_type = relation['infons']['type']\n",
    "\n",
    "            # Finding the texts for the entities in the relations\n",
    "            entity1_text = next((e['text'] for e in all_entities if e['id'] == entity1_id), entity1_id)\n",
    "            entity2_text = next((e['text'] for e in all_entities if e['id'] == entity2_id), entity2_id)\n",
    "\n",
    "            relations.append({'entity1': entity1_text, 'entity2': entity2_text, 'relation': relation_type})\n",
    "\n",
    "        # Adding the same list of relations for the entire document\n",
    "        document_relations.append(relations)\n",
    "\n",
    "    return combined_passages, combined_entities, document_relations\n",
    "\n",
    "# Process the documents\n",
    "passages, entities, relations = process_documents(data['documents'])\n",
    "\n",
    "# Creating the initial DataFrame\n",
    "df_final_combined_with_types = pd.DataFrame({\n",
    "    'passage': passages,\n",
    "    'entities': entities,\n",
    "    'relations': relations\n",
    "})\n",
    "\n",
    "# Function to map entity identifiers to their texts\n",
    "def map_entities_to_texts(entity_list, identifier):\n",
    "    for entity in entity_list:\n",
    "        if entity['identifier'] == identifier:\n",
    "            return entity['text']\n",
    "    return identifier\n",
    "\n",
    "# Modifying the relations to replace identifiers with actual entity texts\n",
    "true_relations = []\n",
    "for index, row in df_final_combined_with_types.iterrows():\n",
    "    modified_relations = []\n",
    "    for relation in row['relations']:\n",
    "        entity1_text = map_entities_to_texts(row['entities'], relation['entity1'])\n",
    "        entity2_text = map_entities_to_texts(row['entities'], relation['entity2'])\n",
    "        modified_relations.append({\n",
    "            'entity1': entity1_text,\n",
    "            'entity2': entity2_text,\n",
    "            'relation': relation['relation']\n",
    "        })\n",
    "    true_relations.append(modified_relations)\n",
    "\n",
    "# Adding the 'true_relations' column to the DataFrame\n",
    "df_final_combined_with_types['true_relations'] = true_relations\n",
    "\n",
    "# Displaying the first data point with the modified 'true_relations' column\n",
    "df_final_combined_with_types.iloc[99].to_dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "413252b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 100 entries, 0 to 99\n",
      "Data columns (total 4 columns):\n",
      " #   Column          Non-Null Count  Dtype \n",
      "---  ------          --------------  ----- \n",
      " 0   passage         100 non-null    object\n",
      " 1   entities        100 non-null    object\n",
      " 2   relations       100 non-null    object\n",
      " 3   true_relations  100 non-null    object\n",
      "dtypes: object(4)\n",
      "memory usage: 3.3+ KB\n"
     ]
    }
   ],
   "source": [
    "df_final_combined_with_types.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fc8340d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>passage</th>\n",
       "      <th>entities</th>\n",
       "      <th>relations</th>\n",
       "      <th>true_relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A novel SCN5A mutation manifests as a malignan...</td>\n",
       "      <td>[{'id': '0', 'text': 'SCN5A', 'type': 'GeneOrG...</td>\n",
       "      <td>[{'entity1': 'D001919', 'entity2': '6331', 're...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity2': 'SCN5A'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Allelic expression imbalance of human mu opioi...</td>\n",
       "      <td>[{'id': '0', 'text': 'human', 'type': 'Organis...</td>\n",
       "      <td>[{'entity1': '4988', 'entity2': 'D010146', 're...</td>\n",
       "      <td>[{'entity1': 'mu opioid receptor', 'entity2': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genetic polymorphisms in the carbonyl reductas...</td>\n",
       "      <td>[{'id': '0', 'text': 'carbonyl reductase 3', '...</td>\n",
       "      <td>[{'entity1': 'D066126', 'entity2': 'C010013', ...</td>\n",
       "      <td>[{'entity1': 'cardiotoxic', 'entity2': 'doxoru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Debrisoquine phenotype and the pharmacokinetic...</td>\n",
       "      <td>[{'id': '0', 'text': 'Debrisoquine', 'type': '...</td>\n",
       "      <td>[{'entity1': '153', 'entity2': 'D008790', 'rel...</td>\n",
       "      <td>[{'entity1': 'beta-1 adrenoceptor', 'entity2':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>The first founder DGUOK mutation associated wi...</td>\n",
       "      <td>[{'id': '0', 'text': 'DGUOK', 'type': 'GeneOrG...</td>\n",
       "      <td>[{'entity1': '1716', 'entity2': 'C580039', 're...</td>\n",
       "      <td>[{'entity1': 'DGUOK', 'entity2': 'hepatocerebr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             passage  \\\n",
       "0  A novel SCN5A mutation manifests as a malignan...   \n",
       "1  Allelic expression imbalance of human mu opioi...   \n",
       "2  Genetic polymorphisms in the carbonyl reductas...   \n",
       "3  Debrisoquine phenotype and the pharmacokinetic...   \n",
       "4  The first founder DGUOK mutation associated wi...   \n",
       "\n",
       "                                            entities  \\\n",
       "0  [{'id': '0', 'text': 'SCN5A', 'type': 'GeneOrG...   \n",
       "1  [{'id': '0', 'text': 'human', 'type': 'Organis...   \n",
       "2  [{'id': '0', 'text': 'carbonyl reductase 3', '...   \n",
       "3  [{'id': '0', 'text': 'Debrisoquine', 'type': '...   \n",
       "4  [{'id': '0', 'text': 'DGUOK', 'type': 'GeneOrG...   \n",
       "\n",
       "                                           relations  \\\n",
       "0  [{'entity1': 'D001919', 'entity2': '6331', 're...   \n",
       "1  [{'entity1': '4988', 'entity2': 'D010146', 're...   \n",
       "2  [{'entity1': 'D066126', 'entity2': 'C010013', ...   \n",
       "3  [{'entity1': '153', 'entity2': 'D008790', 'rel...   \n",
       "4  [{'entity1': '1716', 'entity2': 'C580039', 're...   \n",
       "\n",
       "                                      true_relations  \n",
       "0  [{'entity1': 'bradycardia', 'entity2': 'SCN5A'...  \n",
       "1  [{'entity1': 'mu opioid receptor', 'entity2': ...  \n",
       "2  [{'entity1': 'cardiotoxic', 'entity2': 'doxoru...  \n",
       "3  [{'entity1': 'beta-1 adrenoceptor', 'entity2':...  \n",
       "4  [{'entity1': 'DGUOK', 'entity2': 'hepatocerebr...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_combined_with_types.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a86ea49c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>passage</th>\n",
       "      <th>list_of_entities</th>\n",
       "      <th>true_relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A novel SCN5A mutation manifests as a malignan...</td>\n",
       "      <td>[SCN5A, long QT syndrome, tachycardia, bradyca...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity2': 'SCN5A'...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             passage  \\\n",
       "0  A novel SCN5A mutation manifests as a malignan...   \n",
       "\n",
       "                                    list_of_entities  \\\n",
       "0  [SCN5A, long QT syndrome, tachycardia, bradyca...   \n",
       "\n",
       "                                      true_relations  \n",
       "0  [{'entity1': 'bradycardia', 'entity2': 'SCN5A'...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Re-importing pandas as the code execution state was reset\n",
    "import pandas as pd\n",
    "\n",
    "# Assuming df_final_combined_with_types is already defined and loaded\n",
    "# If not, it should be reloaded or redefined here\n",
    "\n",
    "# Code to create df_for_prompting from df_final_combined_with_types\n",
    "\n",
    "# Extracting the required columns from df_final_combined_with_types\n",
    "passages = df_final_combined_with_types['passage']\n",
    "true_relations = df_final_combined_with_types['true_relations']\n",
    "\n",
    "# Creating list_of_entities column by extracting only the entity text from entities\n",
    "list_of_entities = df_final_combined_with_types['entities'].apply(lambda entities: [entity['text'] for entity in entities])\n",
    "\n",
    "# Creating the new DataFrame df_for_prompting\n",
    "df_for_prompting = pd.DataFrame({\n",
    "    'passage': passages,\n",
    "    'list_of_entities': list_of_entities,\n",
    "    'true_relations': true_relations\n",
    "})\n",
    "\n",
    "# Displaying the first data point of df_for_prompting for verification\n",
    "df_for_prompting.head(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b9e0cb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.Model_Parameter import get_model_response\n",
    "\n",
    "import re\n",
    "import os\n",
    "import csv\n",
    "import json\n",
    "from typing import List\n",
    "import yaml\n",
    "import openai\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from bs4 import BeautifulSoup\n",
    "from tqdm.notebook import tqdm, trange\n",
    "from langchain import PromptTemplate\n",
    "from langchain.prompts.few_shot import FewShotPromptTemplate\n",
    "from langchain.output_parsers import CommaSeparatedListOutputParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4c97aa93",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package punkt to /Users/sanket/nltk_data...\n",
      "[nltk_data]   Package punkt is already up-to-date!\n",
      "[nltk_data] Downloading package wordnet to /Users/sanket/nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "from nltk.translate.bleu_score import corpus_bleu\n",
    "from rouge_score import rouge_scorer\n",
    "import nltk\n",
    "from nltk.translate import meteor_score\n",
    "from nltk.tokenize import word_tokenize\n",
    "nltk.download('punkt')  # Download the 'punkt' resource\n",
    "nltk.download('wordnet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "89fc8f7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package omw-1.4 to /Users/sanket/nltk_data...\n",
      "[nltk_data]   Package omw-1.4 is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import nltk\n",
    "nltk.download('omw-1.4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "dd58943b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= df_for_prompting.head(2).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "cfe3549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>passage</th>\n",
       "      <th>list_of_entities</th>\n",
       "      <th>true_relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A novel SCN5A mutation manifests as a malignan...</td>\n",
       "      <td>[SCN5A, long QT syndrome, tachycardia, bradyca...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity2': 'SCN5A'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Allelic expression imbalance of human mu opioi...</td>\n",
       "      <td>[human, mu opioid receptor, OPRM1, A118G, mu o...</td>\n",
       "      <td>[{'entity1': 'mu opioid receptor', 'entity2': ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             passage  \\\n",
       "0  A novel SCN5A mutation manifests as a malignan...   \n",
       "1  Allelic expression imbalance of human mu opioi...   \n",
       "\n",
       "                                    list_of_entities  \\\n",
       "0  [SCN5A, long QT syndrome, tachycardia, bradyca...   \n",
       "1  [human, mu opioid receptor, OPRM1, A118G, mu o...   \n",
       "\n",
       "                                      true_relations  \n",
       "0  [{'entity1': 'bradycardia', 'entity2': 'SCN5A'...  \n",
       "1  [{'entity1': 'mu opioid receptor', 'entity2': ...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f3eb69db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|                                                     | 0/2 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input Prompt:\n",
      "\n",
      "    Instruction: As a skilled medical domain expert, you are tasked to analyze the relation between entities from the given passage.\n",
      "    The input will contain a list of all the entities present in the passage. \n",
      "    You should find the possible relation between different pairs of  entities.\n",
      "    The relation can only be from the following list: ['Association', 'Positive_Correlation', 'Bind', 'Negative_Correlation', 'Comparison', 'Conversion', 'Cotreatment', 'Drug_Interaction'].\n",
      "    Your response must only contain the following JSON format:\n",
      "    {[{\n",
      "      \"entity1\" : \"name of the entity\",\n",
      "      \"entity2\" : \"name of the entity\",\n",
      "      \"relation\" : \"correct relation from given options\"},\n",
      "      \n",
      "      {\n",
      "      \"entity1\" : \"name of the entity\",\n",
      "      \"entity2\" : \"name of the entity\",\n",
      "      \"relation\" : \"correct relation from given options\"}]\n",
      "      \n",
      "    }\n",
      "\n",
      "\n",
      "    passsage: A novel SCN5A mutation manifests as a malignant form of long QT syndrome with perinatal onset of tachycardia/bradycardia. OBJECTIVE: Congenital long QT syndrome (LQTS) with in utero onset of the rhythm disturbances is associated with a poor prognosis. In this study we investigated a newborn patient with fetal bradycardia, 2:1 atrioventricular block and ventricular tachycardia soon after birth. METHODS: Mutational analysis and DNA sequencing were conducted in a newborn. The 2:1 atrioventricular block improved to 1:1 conduction only after intravenous lidocaine infusion or a high dose of mexiletine, which also controlled the ventricular tachycardia. RESULTS: A novel, spontaneous LQTS-3 mutation was identified in the transmembrane segment 6 of domain IV of the Na(v)1.5 cardiac sodium channel, with a G-->A substitution at codon 1763, which changed a valine (GTG) to a methionine (ATG). The proband was heterozygous but the mutation was absent in the parents and the sister. Expression of this mutant channel in tsA201 mammalian cells by site-directed mutagenesis revealed a persistent tetrodotoxin-sensitive but lidocaine-resistant current that was associated with a positive shift of the steady-state inactivation curve, steeper activation curve and faster recovery from inactivation. We also found a similar electrophysiological profile for the neighboring V1764M mutant. But, the other neighboring I1762A mutant had no persistent current and was still associated with a positive shift of inactivation. CONCLUSIONS: These findings suggest that the Na(v)1.5/V1763M channel dysfunction and possible neighboring mutants contribute to a persistent inward current due to altered inactivation kinetics and clinically congenital LQTS with perinatal onset of arrhythmias that responded to lidocaine and mexiletine.\n",
      "    \n",
      "    entity list: ['SCN5A', 'long QT syndrome', 'tachycardia', 'bradycardia', 'long QT syndrome', 'LQTS', 'patient', 'bradycardia', 'atrioventricular block', 'ventricular tachycardia', 'atrioventricular block', 'lidocaine', 'mexiletine', 'ventricular tachycardia', 'LQTS', 'Na(v)1.5', 'sodium', 'G-->A substitution at codon 1763', 'valine (GTG) to a methionine (ATG)', 'tsA201', 'tetrodotoxin', 'lidocaine', 'V1764M', 'I1762A', 'Na(v)1.5', 'V1763M', 'LQTS', 'arrhythmias', 'lidocaine', 'mexiletine']\n",
      "\n",
      "    \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      " 50%|██████████████████████                      | 1/2 [14:21<14:21, 861.41s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model Output:\n",
      "\n",
      "Input Prompt:\n",
      "\n",
      "    Instruction: As a skilled medical domain expert, you are tasked to analyze the relation between entities from the given passage.\n",
      "    The input will contain a list of all the entities present in the passage. \n",
      "    You should find the possible relation between different pairs of  entities.\n",
      "    The relation can only be from the following list: ['Association', 'Positive_Correlation', 'Bind', 'Negative_Correlation', 'Comparison', 'Conversion', 'Cotreatment', 'Drug_Interaction'].\n",
      "    Your response must only contain the following JSON format:\n",
      "    {[{\n",
      "      \"entity1\" : \"name of the entity\",\n",
      "      \"entity2\" : \"name of the entity\",\n",
      "      \"relation\" : \"correct relation from given options\"},\n",
      "      \n",
      "      {\n",
      "      \"entity1\" : \"name of the entity\",\n",
      "      \"entity2\" : \"name of the entity\",\n",
      "      \"relation\" : \"correct relation from given options\"}]\n",
      "      \n",
      "    }\n",
      "\n",
      "\n",
      "    passsage: Allelic expression imbalance of human mu opioid receptor (OPRM1) caused by variant A118G. As a primary target for opioid drugs and peptides, the mu opioid receptor (OPRM1) plays a key role in pain perception and addiction. Genetic variants of OPRM1 have been implicated in predisposition to drug addiction, in particular the single nucleotide polymorphism A118G, leading to an N40D substitution, with an allele frequency of 10-32%, and uncertain functions. We have measured allele-specific mRNA expression of OPRM1 in human autopsy brain tissues, using A118G as a marker. In 8 heterozygous samples measured, the A118 mRNA allele was 1.5-2.5-fold more abundant than the G118 allele. Transfection into Chinese hamster ovary cells of a cDNA representing only the coding region of OPRM1, carrying adenosine, guanosine, cytidine, and thymidine in position 118, resulted in 1.5-fold lower mRNA levels only for OPRM1-G118, and more than 10-fold lower OPRM1 protein levels, measured by Western blotting and receptor binding assay. After transfection and inhibition of transcription with actinomycin D, analysis of mRNA turnover failed to reveal differences in mRNA stability between A118 and G118 alleles, indicating a defect in transcription or mRNA maturation. These results indicate that OPRM1-G118 is a functional variant with deleterious effects on both mRNA and protein yield. Clarifying the functional relevance of polymorphisms associated with susceptibility to a complex disorder such as drug addiction provides a foundation for clinical association studies.\n",
      "    \n",
      "    entity list: ['human', 'mu opioid receptor', 'OPRM1', 'A118G', 'mu opioid receptor', 'OPRM1', 'pain', 'OPRM1', 'drug addiction', 'A118G', 'N40D', 'OPRM1', 'human', 'A118G', 'A118', 'G118', 'Chinese hamster', 'OPRM1', 'adenosine, guanosine, cytidine, and thymidine in position 118', 'OPRM1', 'G118', 'OPRM1', 'actinomycin D', 'A118', 'G118', 'OPRM1', 'G118', 'drug addiction']\n",
      "\n",
      "    \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████| 2/2 [14:22<00:00, 431.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model Output:\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#first_n_rows_df = df.head(3).copy()\n",
    "# Create a prompt template\n",
    "import time\n",
    "# Start time\n",
    "start_time = time.time()\n",
    "template = (\n",
    "    \"\"\"\n",
    "    Instruction: As a skilled medical domain expert, you are tasked to analyze the relation between entities from the given passage.\n",
    "    The input will contain a list of all the entities present in the passage. \n",
    "    You should find the possible relation between different pairs of  entities.\n",
    "    The relation can only be from the following list: ['Association', 'Positive_Correlation', 'Bind', 'Negative_Correlation', 'Comparison', 'Conversion', 'Cotreatment', 'Drug_Interaction'].\n",
    "    Your response must only contain the following JSON format:\n",
    "    {{[{{\n",
    "      \"entity1\" : \"name of the entity\",\n",
    "      \"entity2\" : \"name of the entity\",\n",
    "      \"relation\" : \"correct relation from given options\"}},\n",
    "      \n",
    "      {{\n",
    "      \"entity1\" : \"name of the entity\",\n",
    "      \"entity2\" : \"name of the entity\",\n",
    "      \"relation\" : \"correct relation from given options\"}}]\n",
    "      \n",
    "    }}\n",
    "\n",
    "\n",
    "    passsage: {passage}\n",
    "    \n",
    "    entity list: {list_of_entities}\n",
    "\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "from tqdm import tqdm  # Import tqdm library\n",
    "\n",
    "# Assuming you have imported pandas and defined 'first_n_rows_df' and 'template' as described\n",
    "\n",
    "# Create an empty list to store model outputs\n",
    "model_outputs = []\n",
    "\n",
    "# Use tqdm as a wrapper around the iteration for tracking progress\n",
    "for index, row in tqdm(df.iterrows(), total=len(df)):\n",
    "    prompt = template.format(\n",
    "        passage=row['passage'],\n",
    "        list_of_entities=row['list_of_entities']\n",
    "    )\n",
    "    # Print the input prompt\n",
    "    print(\"Input Prompt:\")\n",
    "    print(prompt)\n",
    "\n",
    "    output = get_model_response(prompt)[0][1]  # Assuming the response contains the relation\n",
    "    print(\"Model Output:\")\n",
    "    print(output)\n",
    "    \n",
    "    model_outputs.append(output)\n",
    "\n",
    "# Add a new column 'model_output' to 'first_n_rows_df' containing the model outputs\n",
    "df['model_output'] = model_outputs\n",
    "df.head(15)\n",
    "end_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "01544637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>passage</th>\n",
       "      <th>list_of_entities</th>\n",
       "      <th>true_relations</th>\n",
       "      <th>model_output</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A novel SCN5A mutation manifests as a malignan...</td>\n",
       "      <td>[SCN5A, long QT syndrome, tachycardia, bradyca...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity2': 'SCN5A'...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Allelic expression imbalance of human mu opioi...</td>\n",
       "      <td>[human, mu opioid receptor, OPRM1, A118G, mu o...</td>\n",
       "      <td>[{'entity1': 'mu opioid receptor', 'entity2': ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             passage  \\\n",
       "0  A novel SCN5A mutation manifests as a malignan...   \n",
       "1  Allelic expression imbalance of human mu opioi...   \n",
       "\n",
       "                                    list_of_entities  \\\n",
       "0  [SCN5A, long QT syndrome, tachycardia, bradyca...   \n",
       "1  [human, mu opioid receptor, OPRM1, A118G, mu o...   \n",
       "\n",
       "                                      true_relations model_output  \n",
       "0  [{'entity1': 'bradycardia', 'entity2': 'SCN5A'...               \n",
       "1  [{'entity1': 'mu opioid receptor', 'entity2': ...               "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "603de1d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_parquet('./Results/biored_re_llama2_7b_chat_model_response.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8a5d79b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>prompt</th>\n",
       "      <th>ground_truth</th>\n",
       "      <th>mapped_ground_true</th>\n",
       "      <th>model_response</th>\n",
       "      <th>answer</th>\n",
       "      <th>mapped_answer</th>\n",
       "      <th>time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15485686</td>\n",
       "      <td>\\nInstruction: \\nAs a skilled medical domain e...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity1_identifie...</td>\n",
       "      <td>[{'entity1': 'bradycardia', 'entity1_identifie...</td>\n",
       "      <td>Here are the entity relationships reported in ...</td>\n",
       "      <td>[{'entity1_identifier': '', 'entity2_identifie...</td>\n",
       "      <td>[{'entity1_identifier': '', 'entity2_identifie...</td>\n",
       "      <td>12-04-2023-14-31-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16046395</td>\n",
       "      <td>\\nInstruction: \\nAs a skilled medical domain e...</td>\n",
       "      <td>[{'entity1': 'mu opioid receptor', 'entity1_id...</td>\n",
       "      <td>[{'entity1': 'mu opioid receptor', 'entity1_id...</td>\n",
       "      <td>Here are the entity relations found in the pas...</td>\n",
       "      <td>[{'entity1_identifier': '9606', 'entity2_ident...</td>\n",
       "      <td>[{'entity1_identifier': '9606', 'entity2_ident...</td>\n",
       "      <td>12-04-2023-14-34-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18457324</td>\n",
       "      <td>\\nInstruction: \\nAs a skilled medical domain e...</td>\n",
       "      <td>[{'entity1': 'cardiotoxic', 'entity1_identifie...</td>\n",
       "      <td>[{'entity1': 'cardiotoxic', 'entity1_identifie...</td>\n",
       "      <td>Sure! Here are the entity relations reported b...</td>\n",
       "      <td>[{'entity1_identifier': '', 'entity2_identifie...</td>\n",
       "      <td>[{'entity1_identifier': '', 'entity2_identifie...</td>\n",
       "      <td>12-04-2023-14-36-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1848636</td>\n",
       "      <td>\\nInstruction: \\nAs a skilled medical domain e...</td>\n",
       "      <td>[{'entity1': 'beta-1 adrenoceptor', 'entity1_i...</td>\n",
       "      <td>[{'entity1': 'beta-1 adrenoceptor', 'entity1_i...</td>\n",
       "      <td>Here are the entity relationships reported in ...</td>\n",
       "      <td>[{'entity1_identifier': 'D003647', 'entity2_id...</td>\n",
       "      <td>[{'entity1_identifier': 'D003647', 'entity2_id...</td>\n",
       "      <td>12-04-2023-14-37-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19394258</td>\n",
       "      <td>\\nInstruction: \\nAs a skilled medical domain e...</td>\n",
       "      <td>[{'entity1': 'DGUOK', 'entity1_identifier': '1...</td>\n",
       "      <td>[{'entity1': 'DGUOK', 'entity1_identifier': '1...</td>\n",
       "      <td>Here are the entity relations based on the inf...</td>\n",
       "      <td>[{'entity1_identifier': '1716', 'entity2_ident...</td>\n",
       "      <td>[{'entity1_identifier': '1716', 'entity2_ident...</td>\n",
       "      <td>12-04-2023-14-37-59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  record_id                                             prompt  \\\n",
       "0  15485686  \\nInstruction: \\nAs a skilled medical domain e...   \n",
       "1  16046395  \\nInstruction: \\nAs a skilled medical domain e...   \n",
       "2  18457324  \\nInstruction: \\nAs a skilled medical domain e...   \n",
       "3   1848636  \\nInstruction: \\nAs a skilled medical domain e...   \n",
       "4  19394258  \\nInstruction: \\nAs a skilled medical domain e...   \n",
       "\n",
       "                                        ground_truth  \\\n",
       "0  [{'entity1': 'bradycardia', 'entity1_identifie...   \n",
       "1  [{'entity1': 'mu opioid receptor', 'entity1_id...   \n",
       "2  [{'entity1': 'cardiotoxic', 'entity1_identifie...   \n",
       "3  [{'entity1': 'beta-1 adrenoceptor', 'entity1_i...   \n",
       "4  [{'entity1': 'DGUOK', 'entity1_identifier': '1...   \n",
       "\n",
       "                                  mapped_ground_true  \\\n",
       "0  [{'entity1': 'bradycardia', 'entity1_identifie...   \n",
       "1  [{'entity1': 'mu opioid receptor', 'entity1_id...   \n",
       "2  [{'entity1': 'cardiotoxic', 'entity1_identifie...   \n",
       "3  [{'entity1': 'beta-1 adrenoceptor', 'entity1_i...   \n",
       "4  [{'entity1': 'DGUOK', 'entity1_identifier': '1...   \n",
       "\n",
       "                                      model_response  \\\n",
       "0  Here are the entity relationships reported in ...   \n",
       "1  Here are the entity relations found in the pas...   \n",
       "2  Sure! Here are the entity relations reported b...   \n",
       "3  Here are the entity relationships reported in ...   \n",
       "4  Here are the entity relations based on the inf...   \n",
       "\n",
       "                                              answer  \\\n",
       "0  [{'entity1_identifier': '', 'entity2_identifie...   \n",
       "1  [{'entity1_identifier': '9606', 'entity2_ident...   \n",
       "2  [{'entity1_identifier': '', 'entity2_identifie...   \n",
       "3  [{'entity1_identifier': 'D003647', 'entity2_id...   \n",
       "4  [{'entity1_identifier': '1716', 'entity2_ident...   \n",
       "\n",
       "                                       mapped_answer                 time  \n",
       "0  [{'entity1_identifier': '', 'entity2_identifie...  12-04-2023-14-31-37  \n",
       "1  [{'entity1_identifier': '9606', 'entity2_ident...  12-04-2023-14-34-27  \n",
       "2  [{'entity1_identifier': '', 'entity2_identifie...  12-04-2023-14-36-18  \n",
       "3  [{'entity1_identifier': 'D003647', 'entity2_id...  12-04-2023-14-37-15  \n",
       "4  [{'entity1_identifier': '1716', 'entity2_ident...  12-04-2023-14-37-59  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b8fef153",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([{'entity1': 'interleukin 12B', 'entity1_identifier': '3593', 'entity2': 'hepatitis C virus infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'HCV-infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'Interleukin-12', 'entity1_identifier': '3593', 'entity2': 'HCV-infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL-12', 'entity1_identifier': '3593', 'entity2': 'HCV-infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'HCV infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL-12', 'entity1_identifier': '3593', 'entity2': 'HCV infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'HCV infection', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'HCV-infected', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'HCV genotype 1-infected', 'entity2_identifier': 'D006526', 'relation': 'Association'},\n",
       "       {'entity1': 'patients', 'entity1_identifier': '9606', 'entity2': 'IL12B', 'entity2_identifier': '3593', 'relation': 'Association'},\n",
       "       {'entity1': 'IL12B', 'entity1_identifier': '3593', 'entity2': 'patients', 'entity2_identifier': '9606', 'relation': 'Association'}],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['mapped_answer'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79fc7dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('./Results/biored_re_llama2_7b_chat_model_response.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7daaab91",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
